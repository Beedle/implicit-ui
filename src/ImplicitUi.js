import React from 'react';
import PropTypes from 'prop-types';

import {T9nProvider} from './T9n';

export default function ImplicitUi({components, ui, t9n}) {
  return <T9nProvider value={t9n}>{renderUiElements(components, ui, getBaseContext(ui))}</T9nProvider>;
}

ImplicitUi.propTypes = {
  components: PropTypes.object.isRequired,
  t9n: PropTypes.object,
  ui: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object
  ])
};

ImplicitUi.defaultProps = {
  t9n: null,
  ui: []
};


function renderUiElements(components, uiProp, t9nContext) {
  const ui = prepareUiObject(uiProp);

  if (!components) return null;
  if (!ui) return null;

  return (
    ui.map((element, index) => {
      const UiElement = getElement(components, element);
      const context = getElementContext(element, t9nContext);
      const props = getElementProps(components, element, context);
      const children = getElementChildren(element);

      if (!UiElement) return null;
      if (!children) return <UiElement key={index} {...props} t9nContext={context} />;
      const childElements = renderUiElements(components, children, context);

      return <UiElement key={index} {...props} t9nContext={context}>{childElements}</UiElement>;
    })
  );
}

function getElement(components, element) {
  if (typeof element === 'string') return components[element];
  if (typeof element === 'object') return components[element.name];

  return undefined;
}

function prepareUiObject(ui) {
  const TAB_INDENTATION = 2;

  const errorMessage = `Expected a Component or list of Components, but recived: ${JSON.stringify(ui, null, TAB_INDENTATION)}`;

  if (!ui) return ui;
  if (Array.isArray(ui)) return ui;
  if (typeof ui !== 'object') return logErrorAndReturn(errorMessage, null);
  if (ui.name) return [ui];
  if (ui.children && Array.isArray(ui.children)) return ui.children;

  return logErrorAndReturn(errorMessage, null);
}

function logErrorAndReturn(message, returnValue) {
  console.error(message);

  return returnValue;
}

function getBaseContext(ui) {
  if (typeof ui === 'object' && ui.t9nContext) return ui.t9nContext;

  return '';
}

function getElementContext(element, t9nContext) {
  if (element.t9nContext) return `${t9nContext}.${element.t9nContext}`;

  return t9nContext;
}

function getElementProps(components, element, t9nContext) {
  if (typeof element === 'string') return {};
  if (!element.props) return {};

  const replacedProps = scanPropsForComponentsAndReplace(components, element.props, t9nContext);

  return {...element.props, ...replacedProps};
}

function scanPropsForComponentsAndReplace(components, propObjects, t9nContext) {
  const keys = Object.keys(propObjects);
  const componentProps = {};

  keys.forEach((key) => {
    const propObject = propObjects[key];

    if (propObject === null) return;
    if (typeof propObject !== 'object') return;
    if (Array.isArray(propObject)) return;
    if (!propObject.name) return;

    let component = getElement(components, propObject);

    if (!component) return;

    component = createPropsWrapper({components, Component: component, specObject: propObject, t9nContext});
    componentProps[key] = component;
  });

  return componentProps;
}

function createPropsWrapper({components, Component, specObject, t9nContext}) {
  const context = getElementContext(specObject, t9nContext);

  return function PropertyWrapper() {
    if (!specObject.props && !specObject.children) return <Component t9nContext={context} />;
    if (!specObject.children) return <Component {...specObject.props} t9nContext={context} />;
    const childElements = renderUiElements(components, specObject.children, context);

    return <Component {...specObject.props} t9nContext={context}>{childElements}</Component>;
  };
}

function getElementChildren(element) {
  if (typeof element === 'string') return undefined;
  if (!element.children) return undefined;

  return element.children;
}
