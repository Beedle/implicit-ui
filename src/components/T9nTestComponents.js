import React from 'react';
import PropTypes from 'prop-types';

import {withT9n} from '../T9n';

export function Paragraph({text}) {
  return <p>{text}</p>;
}

Paragraph.propTypes = {text: PropTypes.string.isRequired};

export function ShowContext({children, t9nContext}) {
  if (!children) return <span>{t9nContext}</span>;

  return (
    <div>
      <span>{t9nContext}</span>
      <div>
        {children}
      </div>
    </div>
  );
}

ShowContext.propTypes = {
  t9nContext: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element)
  ])
};

ShowContext.defaultProps = {children: undefined};

export const SimpleTag = withT9n(({tagName}) => <span>{tagName}</span>, {tagName: '$simpleTag'});
