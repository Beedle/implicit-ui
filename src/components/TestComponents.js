import React from 'react';
import PropTypes from 'prop-types';

export function Paragraph({text}) {
  return <p>{text}</p>;
}

Paragraph.propTypes = {text: PropTypes.string.isRequired};

export function MyTitle() {
  return <h1>My Title</h1>;
}

export function Container({children, text}) {
  return (
    <div>
      <h3>{text}</h3>
      <div>
        {children}
      </div>
    </div>
  );
}

Container.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element)
  ]),
  text: PropTypes.string
};

Container.defaultProps = {
  children: <span>Empty Container</span>,
  text: 'Container Header'
};

export function PropsContainer({Comp1, Comp2}) {
  return (
    <div>
      <Comp1 />
      <Comp2 />
    </div>
  );
}

PropsContainer.propTypes = {
  Comp1: PropTypes.func,
  Comp2: PropTypes.func
};

PropsContainer.defaultProps = {
  Comp1: <span>Default1</span>,
  Comp2: <span>Default2</span>
};
