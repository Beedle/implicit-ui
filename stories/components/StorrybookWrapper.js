import React from 'react';

import './StorybookWrapper.scss'; //eslint-disable-line

export default function StorybookWrapper({children}) {
  return (
    <div className="storybook-wrapper">{children}</div>
  );
}
