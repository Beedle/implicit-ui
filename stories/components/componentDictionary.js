import React from 'react';

export function MyParagraph() {
  return <p>This is my Paragraph</p>;
}

export function MyTextParagraph({text}) {
  return <p>{text}</p>;
}

export function Container({children, text}) {
  return (
    <div>
      <h3>{text}</h3>
      <div>
        {children}
      </div>
    </div>
  );
}

export function PropsContainer({Comp1, Comp2}) {
  return (
    <div>
      <Comp1 />
      <Comp2 />
    </div>
  );
}
