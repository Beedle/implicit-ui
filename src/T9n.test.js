import React from 'react';
import renderer from 'react-test-renderer';

import ImplicitUi from './ImplicitUi';
import * as compCollection from './components/T9nTestComponents';

const uiConfig = {
  t9nContext: 'test',
  children: [
    'SimpleTag',
    {
      t9nContext: 'level2',
      name: 'SimpleTag'
    }
  ]
};

describe('T9n', () => {
  it('should use full context and tagname as default if no translation exist', () => {
    const tree = renderer
      .create(<ImplicitUi components={compCollection} showDefaultTag ui={uiConfig} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should use empty string if showDefaultTag is not set', () => {
    const tree = renderer
      .create(<ImplicitUi components={compCollection} ui={uiConfig} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should use full contextTag if exists', () => {
    const t9n = {
      test$simpleTag: 'Translated Simple Tag',
      'test.level2$simpleTag': 'Translated Simple Tag level 2'
    };
    const tree = renderer
      .create(<ImplicitUi components={compCollection} t9n={t9n} ui={uiConfig} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should use fallback to lower context if full context does not exist', () => {
    const t9n = {
      $simpleTag: 'Translated Simple Tag',
      level2$simpleTag: 'Translated Simple Tag level 2'
    };
    const tree = renderer
      .create(<ImplicitUi components={compCollection} t9n={t9n} ui={uiConfig} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should use fallback al the way to the bare tag', () => {
    const t9n = {$simpleTag: 'Translated Simple Tag'};
    const tree = renderer
      .create(<ImplicitUi components={compCollection} t9n={t9n} ui={uiConfig} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should override tagname if tag prop is sent in', () => {
    const uiConfigOveride = {
      t9nContext: 'test',
      children: [
        {
          t9nContext: 'level2',
          name: 'SimpleTag',
          props: {tagName: 'override$differentTag'}
        }
      ]
    };
    const t9n = {$simpleTag: 'Translated Simple Tag'};
    const tree = renderer
      .create(<ImplicitUi components={compCollection} showDefaultTag t9n={t9n} ui={uiConfigOveride} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
