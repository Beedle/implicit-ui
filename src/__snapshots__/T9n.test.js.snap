// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`T9n should override tagname if tag prop is sent in 1`] = `
<span>
  test.level2.override$differentTag
</span>
`;

exports[`T9n should use empty string if showDefaultTag is not set 1`] = `
Array [
  <span>
    
  </span>,
  <span>
    
  </span>,
]
`;

exports[`T9n should use fallback al the way to the bare tag 1`] = `
Array [
  <span>
    Translated Simple Tag
  </span>,
  <span>
    Translated Simple Tag
  </span>,
]
`;

exports[`T9n should use fallback to lower context if full context does not exist 1`] = `
Array [
  <span>
    Translated Simple Tag
  </span>,
  <span>
    Translated Simple Tag level 2
  </span>,
]
`;

exports[`T9n should use full context and tagname as default if no translation exist 1`] = `
Array [
  <span>
    test$simpleTag
  </span>,
  <span>
    test.level2$simpleTag
  </span>,
]
`;

exports[`T9n should use full contextTag if exists 1`] = `
Array [
  <span>
    Translated Simple Tag
  </span>,
  <span>
    Translated Simple Tag level 2
  </span>,
]
`;
